BosunReporter/Handlers/DataDogStatsdMetricHandler.cs

```diff
@@ -0,0 +1,294 @@
+
+                    // remove the last comma
+                    bytesWritten--;
+                }
+
+                CopyToBuffer(s_newLine);
+
+                // now write it to the buffer writer
+                writer.Write(memory.Span.Slice(0, length));
+
+                void CopyToBuffer(byte[] source)
+                {
+                    Array.Copy(source, 0, buffer, bytesWritten, source.Length);
+                    bytesWritten += source.Length;
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        protected override BufferWriter<byte> CreateBufferWriter(PayloadType payloadType)
+        {
+            switch (payloadType)
+            {
+                case PayloadType.Counter:
+                case PayloadType.CumulativeCounter:
+                case PayloadType.Gauge:
+                    return _metricBufferWriter.Value;
+                case PayloadType.Metadata:
+                    return _metadataBufferWriter.Value;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(payloadType));
+            }
+        }
+
+        private ValueTask<IPEndPoint> CreateEndpointAsync(string host, ushort port)
+        {
+            if (IPAddress.TryParse(host, out IPAddress ip))
+            {
+                return new ValueTask<IPEndPoint>(new IPEndPoint(ip, port));
+            }
+
+            return GetEndpointAsync();
+
+            async ValueTask<IPEndPoint> GetEndpointAsync()
+            {
+                var hostAddresses = await Dns.GetHostAddressesAsync(host);
+                var hostAddress = hostAddresses.FirstOrDefault(x => x.AddressFamily == AddressFamily.InterNetwork || x.AddressFamily == AddressFamily.InterNetworkV6);
+                if (hostAddress == null)
+                {
+                    throw new ArgumentException("Unable to find an IPv4 or IPv6 address for host", nameof(host));
+                }
+
+                if (hostAddress.AddressFamily == AddressFamily.InterNetworkV6)
+                {
+                    return new IPEndPoint(hostAddress.MapToIPv6(), port);
+                }
+
+                return new IPEndPoint(hostAddress.MapToIPv4(), port);
+            }
+        }
+
+        private ValueTask SendMetricAsync(PayloadType type, ReadOnlyMemory<byte> buffer)
+        {
+            if (!MemoryMarshal.TryGetArray(buffer, out var arraySegment))
+            {
+                throw new InvalidOperationException("Could not retrieve underlying array from ReadOnlyMemory<byte>");
+            }
+
+            var endpointTask = _endpointFactory.Value;
+            if (endpointTask.IsCompleted)
+            {
+                var sendTask = Task.Factory.FromAsync(
+                    _socket.BeginSendTo(arraySegment.Array, arraySegment.Offset, arraySegment.Count, SocketFlags.None, endpointTask.Result, null, _socket),
+                    _socket.EndSendTo
+                );
+
+                if (sendTask.IsCompleted)
```

Comment:
would we *ever* expect this to be already completed? not sure we would...

any reason not to like the AsyncSocketEventArgs API here?

Reply:
btw; if you need AsyncSocketEventArgs to be awaitable, I have "bits" for that

Reply:
hey, did I mention that I have an incomplete spike that is a high perf UDP wrapper designed to offload network layers to the lib?